pennylane/templates/subroutines/prepselprep.py

```diff
@@ -0,0 +1,144 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the PrepSelPrep template.
+This template contains a decomposition for performing a block-encoding on a
+linear combination of unitaries using the Prepare, Select, Prepare method.
+"""
+# pylint: disable=arguments-differ
+import copy
+
+import pennylane as qml
+from pennylane.operation import Operation
+
+
+class PrepSelPrep(Operation):
+    """This class implements a block-encoding of a linear combination of unitaries
+    using the Prepare, Select, Prepare method"""
+
+    def __init__(self, lcu, control, id=None):
+        coeffs, ops = lcu.terms()
+        control = qml.wires.Wires(control)
+        self.hyperparameters["lcu"] = lcu
+        self.hyperparameters["coeffs"] = coeffs
+        self.hyperparameters["ops"] = tuple(ops)
+        self.hyperparameters["control"] = control
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+        target_wires = qml.wires.Wires.all_wires([op.wires for op in ops])
+        self.hyperparameters["target_wires"] = target_wires
+
+        all_wires = target_wires + control
+        super().__init__(*self.data, wires=all_wires, id=id)
+
+    def _flatten(self):
+        return tuple(self.lcu), (self.control)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "PrepSelPrep":
+        coeffs = [term.terms()[0][0] for term in data]
+        ops = [term.terms()[1][0] for term in data]
+        lcu = qml.ops.LinearCombination(coeffs, ops)
+        return cls(lcu, metadata)
+
+    def __repr__(self):
+        return f"PrepSelPrep(coeffs={tuple(self.coeffs)}, ops={self.ops}, control={self.control})"
+
+    def map_wires(self, wire_map: dict) -> "Select":
+        new_ops = [o.map_wires(wire_map) for o in self.hyperparameters["ops"]]
+        new_control = [wire_map.get(wire, wire) for wire in self.hyperparameters["control"]]
+        new_lcu = qml.dot(self.hyperparameters["coeffs"], new_ops)
+        return PrepSelPrep(new_lcu, new_control)
+
+    def decomposition(self):
+        return self.compute_decomposition(self.lcu, self.control)
+
+    @staticmethod
+    def compute_decomposition(lcu, control):
+        coeffs, ops = lcu.terms()
+        normalized_coeffs = qml.math.sqrt(qml.math.abs(coeffs)) / qml.math.norm(
+            qml.math.sqrt(qml.math.abs(coeffs))
+        )
+
+        with qml.QueuingManager.stop_recording():
+            prep_ops = qml.StatePrep.compute_decomposition(normalized_coeffs, control)
+            select_ops = qml.Select.compute_decomposition(ops, control)
+            adjoint_prep_ops = list(map(qml.adjoint, prep_ops))
+
+        ops = prep_ops + select_ops + adjoint_prep_ops
```

Comment:
Something here seems wrong to me, could you explain what you are trying to do?

Specifically, I would bet that the decomposition you generate for the case where `prep_ops = [RZ(1.23, 'aux0'), RX(4.5, 'aux0')], select_ops = [X(0), Y(0)]` is incorrect.

Reply:
I think this is the problem. Mapping `qml.adjoint` over `prep_ops` is not actually computing the adjoint since Adjoint(AB) = Adjoint(B)Adjoint(A). I think this is a better way to do it:

`adjoint_prep_ops = qml.adjoint(qml.StatePrep(normalized_coeffs, control)).decomposition()`

When I wrote the original code I was less familiar with the Operator class and didn't realize I could just call decomposition on an Adjoint. My tests pass with this update. Could you suggest an LCU to test that produces an example like the one you gave?

Reply:
Yes, you are correct! Actually it turns out that the reason your code was passing to begin with was that the decomposition of `StatePrep` is just a single operation called `MottonenStatePreparation` which itself has a decomposition. 

It turns out that PennyLane will recursively decompose operations that it can natively execute on hardware. What this means is that you don't actually need to "manually" decompose the prep, select and prep^dag operations and apply them. 

you can instead do something *like this* (note this is not the full solution):

```
class PrepSelPrep:
    
    def compute_decomposition(coeffs, ops):
        decomp = [qml.StatePrep(...), qml.Select(...), qml.adjoint(qml.StatePrep)]
        return decomp
```

Reply:
I see, my old version queuing the next level of the recursion. This is why I was seeing ```MottonenStatePreparation``` in the queue instead of ``StatePrep```. I changed my code based on your example, and updated the tests. It's a lot cleaner now.